src/controllers/users.controller.js

```diff
@@ -263,6 +267,127 @@ class UserController {
     const allUsers = await db.user.findAll(options);
     return Responses.handleSuccess(200, 'successfully retrieved all users', res, allUsers);
   }
+
+  /**
+   * Upload travel documents
+   * @param {object} req
+   * @param {object} res
+   * @returns {object} response
+   */
+  async uploadDocument(req, res) {
+    const {
+      name,
+    } = req.body;
+
+    let url = '';
+    if (req.file !== undefined) {
+      const {
+        filename,
+        path,
+        mimetype
+      } = req.file;
+
+      url = await filesService.s3Upload({ path, filename, mimetype }, 'documents');
+    }
+
+    const { userId } = res.locals.user;
+
+    const document = await UserService.addDocument({ name, url, userId });
+    Responses.handleSuccess(201, 'Document uploaded successfully', res, { name, url, userId });
+  }
+
+  /**
+   * Delete travel document
+   * @param {object} req
+   * @param {object} res
+   * @returns {object} response
+   */
+  async deleteDocument(req, res) {
+    const { id } = req.params;
+    const { userId } = res.locals.user;
+
+    const document = await UserService.getDocument(id);
+
+    const isOwner = document.userId === userId;
+
+    if (isOwner) {
+      await UserService.deleteDocument(id);
+      return Responses.handleSuccess(200, 'Document Deleted successfully', res);
+    }
+    return Responses.handleError(409, 'Document doesn\'t belong to you', res);
+  }
+
+
+  /**
+   * Retrieve travel document
+   * @param {object} req
+   * @param {object} res
+   * @returns {object} response
+   */
+  async getDocuments(req, res) {
+    const { userId, role } = res.locals.user;
+
+    const documents = await UserService.retrieveDocuments({
+      ...role === 'requester' && { userId },
+      ...role === 'travel_administrator' && { travelAdminId: userId },
+      ...role === 'suppliers' && { travelAdminId: userId },
+    });
+
+    Responses.handleSuccess(200, 'Documents retrieved successfully', res, documents);
+  }
+
+  /**
+   * verify a travel document
+   * @param {object} req
+   * @param {object} res
+   * @returns {object} response
+   */
+  async verifyDocument(req, res) {
+    const { id } = req.params;
+    const { userId } = res.locals.user;
+    await UserService.verifyDocument(id, userId);
+    return Responses.handleSuccess(200, 'Document verified successfully', res);
+  }
+
+  /**
+   * Preview document
+   * @param {Object} req
+   * @param {Object} res
+   * @returns {File} document
+   */
+  async downloadDocument(req, res) {
+    const { id } = req.params;
+    const document = await UserService.getDocument(id);
+
+    const dir = 'public/temp';
+
+    if (!fs.existsSync(dir)) {
+      fs.mkdirSync(dir);
+    }
+    const filename = `${document.documentOwner.firstName}-`
+    + `${document.documentOwner.lastName}-${document.name}`
+    + `.${document.url.split('.').pop()}`;
```

Comment:
Separation of concerns, You need to thoroughly test this part of the code. I am very sure it can fail in so many scenarios.

Reply:
Going to refactor this too

Reply:
And here is what I did, what do you think @kipropbrian?
```
const documentName = `${document.name}`;
    const documentOwner = `${document.documentOwner.firstName}-${document.documentOwner.lastName}`;
    const fileExtension = `${document.url.split('.').pop()}`;

    const filename = `${documentOwner}-${documentName}-${(new Date()).getTime()}.${fileExtension}`;
    const file = fs.createWriteStream(`public/temp/${filename}`);
```